packages/parser/src/parse/alias.ts

```diff
@@ -1,302 +1,369 @@
+  type TokenNormalized,
+  type TokenNormalizedSet,
+  type TransitionTokenNormalized,
+  type TypographyTokenNormalized,
+  isAlias,
+  parseAlias,
+} from '@terrazzo/token-tools';
 import type Logger from '../logger.js';
+import { getObjMembers } from './json.js';
 
-/** Throw error if resolved alias for composite properties doesn’t match expected $type. */
-export const COMPOSITE_TYPE_VALUES = {
-  border: {
-    color: ['color'],
-    width: ['dimension'],
-    strokeStyle: ['strokeStyle'],
-  },
-  gradient: {
-    color: ['color'],
-    position: ['number'],
-  },
-  shadow: {
-    color: ['color'],
-    position: ['dimension'],
-  },
-  strokeStyle: {
-    dashArray: ['dimension'],
-  },
-  transition: {
-    duration: ['duration'],
-    delay: ['duration'],
-    timingFunction: ['cubicBezier'],
-  },
-  typography: {
-    fontFamily: ['fontFamily'],
-    fontSize: ['dimension'],
-    fontWeight: ['fontWeight'],
-    letterSpacing: ['dimension'],
-    lineHeight: ['dimension', 'number'],
-  },
+export interface ApplyAliasOptions {
+  tokensSet: TokenNormalizedSet;
+  filename: URL;
+  src: string;
+  node: ObjectNode;
+  logger: Logger;
+}
+
+export type PreAliased<T extends TokenNormalized> = {
+  $value: T['$value'] | string;
+  mode: Record<string, T['mode'][string] & { $value: T['$value'] | string }>;
 };
 
 /**
- * Resolve alias
- * Note: to save work during resolution (and reduce memory), this will inject
- * `aliasedBy` while it traverses. Resolution isn’t normally associated with
- * mutation, but for our usecase it is preferable.
+ * Resolve aliases and update the token nodes.
+ *
+ * Data structures are in an awkward in-between phase, where they have
+ * placeholders for data but we still need to resolve everything. As such,
+ * TypeScript will raise errors expecting the final shape.
+ *
+ * This is also a bit tricky because different token types alias slightly
+ * differently. For example, color tokens and other “primitive” tokens behave
+ * as-expected. But composite tokens like Typography, Gradient, Border, etc. can
+ * either fully- or partially-alias their values. Then we add modes to the mix,
+ * and we have to do the work all over again for each mode declared.
+ *
+ * All that to say, there are a generous amount of TypeScript overrides here rather
+ * than try to codify indeterminate shapes.
  */
-export function resolveAlias(
-  alias: string,
-  {
-    tokens,
-    logger,
-    filename,
-    src,
-    node,
-    scanned = [],
-  }: {
-    tokens: Record<string, TokenNormalized>;
-    logger: Logger;
-    filename?: URL;
-    src: string;
-    node: AnyNode;
-    scanned?: string[];
-  },
-): { id: string; chain: string[] } {
-  const { id } = parseAlias(alias);
-  if (!tokens[id]) {
-    logger.error({ group: 'parser', label: 'alias', message: `Alias "${alias}" not found.`, filename, src, node });
+export default function applyAliases(token: TokenNormalized, options: ApplyAliasOptions): void {
```

Comment:
This rewrite made me want to pull my hair out. But here I am, (most) hairs in tact